jets-bench/src/input.rs

```diff
@@ -93,3 +441,934 @@ impl InputSampling {
+pub struct VarLengthBufferExact<const N: usize>;
+
+impl InputSample for VarLengthBuffer {
+    fn n_distributions(&self) -> usize {
+        2
+    }
+
+    fn distribution_name(&self, dist: usize) -> String {
+        match dist {
+            0 => "varbuf-rndlen".into(),
+            1 => "varbuf-maxlen".into(),
+            x => panic!("no distribution {x} for VarLengthBuffer"),
+        }
+    }
+
+    fn sample(&self, dist: usize, n_bits: usize) -> FlatValue {
+        let mut n = None;
+        for i in 0..16 {
+            if n_bits == (8 << (i + 1)) - 8 + i + 1 {
+                n = Some(i);
+            }
+        }
+        let n = match n {
+            Some(n) => n,
+            None => panic!(
+                "n_bits must be of the form 2^n + n, not {} for n in 0..16",
+                n_bits,
+            ),
+        };
+
+        let mut rng = rand::thread_rng();
+
+        let max_value_len_bytes = 1 << (n + 1);
+        let value_len_bytes = match dist {
+            // uniform
+            0 => rng.gen_range(0..=max_value_len_bytes),
+            1 => max_value_len_bytes,
+            x => panic!("no distribution {x} for VarLengthBuffer"),
+        };
+
+        // Construct the value *as put on the bit machine*, which means
+        // that the units show up as long strings of 0s rather than just
+        // a single 0 bit (as would be done for the witness encoding).
+        let mut ret = FlatValue::zero_n_bits(0);
+        for i in (0..n + 1).rev() {
+            let rhs = if value_len_bytes & (1 << i) == 0 {
+                FlatValue::zero_n_bits(8 << i).prefix_zero()
+            } else {
+                FlatValue::random_n_bits(8 << i).prefix_one()
+            };
+            ret = FlatValue::product(&[ret, rhs]);
+        }
+        assert_eq!(ret.len_bits, n_bits); // sanity check
+        ret
+    }
+}
+
+impl<const N: usize> InputSample for VarLengthBufferExact<N> {
+    fn n_distributions(&self) -> usize {
+        VarLengthBuffer.n_distributions()
+    }
+
+    fn distribution_name(&self, dist: usize) -> String {
+        VarLengthBuffer.distribution_name(dist)
+    }
+
+    fn sample(&self, dist: usize, n_bits: usize) -> FlatValue {
+        assert_eq!(
+            n_bits,
+            self.n_bits(),
+            "requested an {}-bit sample from a varlength buffer with N = {} (len = {})",
+            n_bits,
+            N,
+            self.n_bits(),
+        );
+        VarLengthBuffer.sample(dist, n_bits)
+    }
+}
+impl<const N: usize> InputSampleExactSize for VarLengthBufferExact<N> {
+    fn n_bits(&self) -> usize {
+        (8 << (N + 1)) - 8 + N + 1
+    }
+}
+
+pub struct EqProduct<L>(pub L);
+
+impl<L: InputSample> InputSample for EqProduct<L> {
+    fn n_distributions(&self) -> usize {
+        // both random
+        self.0.n_distributions() * self.0.n_distributions()
+            // both equal, and each a bit-inverse of the other
+            + 2 * self.0.n_distributions()
```

Comment:
1eb9db38d89add11c740e93ec15a21793ce34945
> `n + n + 2 * n`

I'm just skimming this code, but I expected `2n` instead of `4n`.

Reply:
I don't know what to make of this comment. `GenericProduct` does `n*m` as you would expect and `EqProduct` can do other distributions because it knows both factors have the same size, so it won't behave the same way as `EqProduct<L>`.

I don't see `4n` anywhere in the codebase.

Reply:
For `EqProduct<L>` I read `self.0.n_distributions() * self.0.n_distributions() + 2 * self.0.n_distributions()` as `4 * self.0.n_distributions()`. For `GenericProduct<L, L>` (with the same generic parameter twice) I read `self.0.n_distributions() * self.1.n_distributions()` as `2 * self.0.n_distributions()`. This is of course wrong, but I hope you see where I was coming from.